lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+ * @param canport FDCAN block base address
+ * @returns Non-negative number ID of Tx buffer which is free,
+ * or FDCAN_E_BUSY if no Tx buffer is available
+ */
+static int fdcan_get_free_txbuf(uint32_t canport)
+{
+	if ((FDCAN_TXBRP(canport) & FDCAN_TXBRP_TRP0) == 0)
+	{
+		return 0;
+	} 
+	else if ((FDCAN_TXBRP(canport) & FDCAN_TXBRP_TRP1) == 0)
+	{
+		return 1;
+	} 
+	else if ((FDCAN_TXBRP(canport) & FDCAN_TXBRP_TRP2) == 0)
+	{
+		return 2;
+	} 
+
+	return FDCAN_E_BUSY;
+}
+
+/** Returns fill state and next available get index from receive FIFO.
+ * Examines FDCAN receive FIFO and returns fill status of FIFO and ID of
+ * next message available for reading. If fill status is 0 (FIFO is empty), 
+ * then get index is undefined.
+ * @param canport FDCAN block base address
+ * @param fifo_id ID of fifo queried (currently 0 or 1)
+ * @param get_index Address of buffer where next get index will be stored
+ * @param pending_frames Address of buffer where amount of pending frame will be stored.
+ * @returns FDCAN_E_OK if buffers were filled with data, FDCAN_E_INVALID if
+ * FIFO ID is invalid.
+ */
+static int fdcan_get_fill_rxfifo(uint32_t canport, uint8_t fifo_id, unsigned * get_index, unsigned * pending_frames)
+{
+	if (fifo_id == 0)
+	{
+		*get_index = (FDCAN_RXF0S(canport) & FDCAN_RXF0S_F0GI_MASK) 
+			>> FDCAN_RXF0S_F0GI_SHIFT;
+		*pending_frames = (FDCAN_RXF0S(canport) & FDCAN_RXF0S_F0FL_MASK) 
+			>> FDCAN_RXF0S_F0FL_SHIFT;
+	}
+	else if (fifo_id == 1)
+	{
+		*get_index = (FDCAN_RXF1S(canport) & FDCAN_RXF1S_F1GI_MASK) 
+					>> FDCAN_RXF1S_F1GI_SHIFT;
+		*pending_frames = (FDCAN_RXF1S(canport) & FDCAN_RXF1S_F1FL_MASK) 
+			>> FDCAN_RXF1S_F1FL_SHIFT;
+	}
+	else
+	{
+		return FDCAN_E_INVALID;
+	}
+
+	return FDCAN_E_OK;
+}
+
+/** Obtain address of FDCAN Message RAM for certain FDCAN block.
+ * @param canport identification of FDCAN block. Use CANx defines here.
+ * @return Address of Message RAM for given FDCAN block or null pointer
+ * if FDCAN block identification is invalid.
+ */
+static struct FDCAN_MessageRAM * fdcan_get_msgram_addr(uint32_t canport)
+{
+	/* This piece of code may look wrong, after one examines
+	 * STM32G4 datasheet and/or g4/memorymap.h. There are three
+	 * memory regions defined for FDCANx_RAM_BASE. They are 0x400
+	 * bytes apart as per chapter 2.2.2 of [RM0440].
+	 *
+	 * It turns out, that these addresses are not in line with what 
+	 * is specified later in chapter 44.3.3 of [RM0440]. There it is 
+	 * stated, that message RAMs are packed and in case of multiple 
+	 * FDCAN blocks, message RAM for n-th FDCAN starts at address 
+	 * end of (n-1)-th block + 4 (explicitly, offset 0x354).
+	 *
+	 * It turns out, that this statement is also false! In fact FDCAN
+	 * message RAMs are packed tightly and n-th block starts immediately
+	 * after (n-1)-th block ends. Thus offset is going to be computed
+	 * using formula:
+	 *
+	 * FDCAN1_RAM_BASE + (block_id * sizeof(struct FDCAN_MessageRAM))
+	 */
```

Comment:
why is this not just 3 constants?

Reply:
In certain controllers, such as H7, size of Message RAM per CPU block is configurable. Constants probably wouldn't work well here. I originally expected that different MCU families will have different fixed sizes, but user-configurable size calls for completely different approach.

Reply:
ok, but this function is constant, there's no support for configurable message ram here?

Reply:
No, as this is written and tested, it only supports constant mapping CANFDx_BASE => CANFDx_MSGRAM_BASE. Code below should end up as returning constants as all value are compile-time deducible.